### PR TITLE
add router replay to latentMoE models

### PR DESCRIPTION
**Pull request by faresobeid**
Previously was a no-op

---

> [!NOTE]
> **Medium Risk**
> Changes MoE routing behavior on the training path when `routed_experts` is supplied, which can affect gradients and load-balancing stats; default forward path is unchanged when the tensor is omitted.
> 
> **Overview**
> Implements **router replay** for NemotronH **LatentMoE** so training can match inference expert choices instead of treating `routed_experts` as a no-op.
> 
> **`NemotronHRouter`** gains an optional `routed_experts` argument. When present, it **reuses the supplied expert indices** (from the inference engine) and **only recomputes top-k gating weights** from the trainer’s sigmoid scores; group routing, `e_score_correction_bias`, and load-balancing `expert_bias` are **skipped** for selection (they only affect which experts are picked in the normal path).
> 
> **`LatentMoE`** flattens per-token `routed_experts` and passes them into the router. **`NemotronHModel` / `NemotronHForCausalLM`** accept `routed_experts` with shape `(batch, seq, num_hidden_layers, top_k)`, slice the tensor **per global layer index** in the decoder loop, and thread the slice through Mamba, attention, and MoE layers (non-MoE layers ignore it).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 1475afd569222c589cf0384b2f74a6f916c5ed77. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/prime_rl/trainer/models/layers/moe.py [hunk 1]
```diff
@@ -955,32 +955,44 @@ def __init__(
...
-            )
-            group_idx = torch.topk(group_scores, k=self.topk_group, dim=-1, sorted=False)[1]
-            group_mask = torch.zeros_like(group_scores)
-            group_mask.scatter_(1, group_idx, 1)
-            score_mask = (
-                group_mask.unsqueeze(-1)
-                .expand(-1, self.n_group, self.num_experts // self.n_group)
-                .reshape(-1, self.num_experts)
-            )
-            scores_for_choice = scores_for_choice.masked_fill(~score_mask.bool(), 0.0)
```
Comment (samsja) [inline src/prime_rl/trainer/models/layers/moe.py:981]: wut we ddnt had r3 here ? how is it even working

Comment (faresobeid) [reply src/prime_rl/trainer/models/layers/moe.py:981]: ya idk tbh i guess it somehow no-ops

Comment (samsja) [reply src/prime_rl/trainer/models/layers/moe.py:981]: hmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmm i stil dont get it

Comment (faresobeid) [reply src/prime_rl/trainer/models/layers/moe.py:981]: ya it just landed in kwargs and ignored

Comment (S1ro1) [reply src/prime_rl/trainer/models/layers/moe.py:981]: hmmm i wonder who implemented nemotron and if it was after R3 or before (it might as well have been before and Id idn't do it LOL)

**Review by samsja — APPROVED** [review]
